src/main/java/im/status/keycard/SecureChannel.java

```diff
@@ -207,33 +284,60 @@ public void pair(APDU apdu) {
+   * Performs the first step of certificate based pairing. In this step, the card provides a CA signed certificate
+   * of its card ID key, and proves ownership of this key with a signature on a challenge hash. The challenge hash
+   * to be signed is computed as the sha256 hash of the client salt (provided by client) and card salt (generated randomly).
+   * The card will also include its random salt in the response to the client, so that the client can reproduce the challenge
+   * hash and verify the card signature.
    *
    * @param apduBuffer the APDU buffer
    * @return the length of the reply
    */
   private short pairStep1(byte[] apduBuffer) {
-    preassignedPairingOffset = -1;
-
-    for (short i = 0; i < (short) pairingKeys.length; i += PAIRING_KEY_LENGTH) {
-      if (pairingKeys[i] == 0) {
-        preassignedPairingOffset = i;
-        break;
-      }
+    // Make sure certificate exisits
+    if (idCertStatus != ID_CERTIFICATE_LOCKED) {
+      ISOException.throwIt(ISO7816.SW_CONDITIONS_NOT_SATISFIED);
     }
 
-    if (preassignedPairingOffset == -1) {
-      ISOException.throwIt(ISO7816.SW_FILE_FULL);
+    // Override first pairing slot
+    preassignedPairingOffset = 0;
+
+    // Compute ECDH secret
+    crypto.ecdh.init(idKeypair.getPrivate());
+    short pubKeyOff = (short)(ISO7816.OFFSET_CDATA + SC_SECRET_LENGTH);
+    try {
+      crypto.ecdh.generateSecret(apduBuffer, (short) (pubKeyOff + 1), apduBuffer[pubKeyOff], secret, (short) 0);
+    } catch(Exception e) {
+      ISOException.throwIt(ISO7816.SW_WRONG_DATA);
+      return 0;
     }
 
-    crypto.sha256.update(pairingSecret, (short) 0, SC_SECRET_LENGTH);
-    crypto.sha256.doFinal(apduBuffer, ISO7816.OFFSET_CDATA, SC_SECRET_LENGTH, apduBuffer, (short) 0);
-    crypto.random.generateData(secret, (short) 0, SC_SECRET_LENGTH);
-    Util.arrayCopyNonAtomic(secret, (short) 0, apduBuffer, SC_SECRET_LENGTH, SC_SECRET_LENGTH);
+    // Hash client salt and ECDH secret and store in secret buffer
+    // secretHash = sha256(clientSalt, ECDH secret)
+    crypto.sha256.update(apduBuffer, ISO7816.OFFSET_CDATA, SC_SECRET_LENGTH);
+    crypto.sha256.doFinal(secret, (short) 0, SC_SECRET_LENGTH, secret, (short) 0);
+
+    // Generate random card salt, and copy to response buffer
+    crypto.random.generateData(apduBuffer, (short) 0, SC_SECRET_LENGTH);
+
+    // Copy card certificate to response buffer
+    Util.arrayCopyNonAtomic(idCertificate, (short) 0, apduBuffer, SC_SECRET_LENGTH, (short) CERTIFICATE_LEN);
+
+    // Sign the secret hash, and copy the signature into the response buffer
+    eccSig.init(idKeypair.getPrivate(), Signature.MODE_SIGN);
+    short len = eccSig.signPreComputedHash(secret, (short) 0, (short) (SC_SECRET_LENGTH), apduBuffer, (short) (SC_SECRET_LENGTH + CERTIFICATE_LEN));
 
-    return (SC_SECRET_LENGTH * 2);
+    // Zero-pad end of signature field (ECDSA sig is variable length, but we send full max size field)
```

Comment:
Why? DER specifies the length and this seems unnecessary, dangerous, and confusing

Reply:
Because everything else coming out of the applet is fixed length and when we (you, Aurash, and I) talked about this, you two promised me that all applet payloads would stay fixed length, so we do not require BER-TLV processing on the client side.

I can BER-TLV encode it, if you'd rather, but that differs from our original discussion

Reply:
I vaguely remember that discussion, but I was not referring to DER signatures. You're already going to have to parse them for Status' signature template -- why not carry over that DER parser to this response too?

Reply:
> why not carry over that DER parser to this response too?

Well we hadn't implemented that command yet, and I was told that all commands would return fixed-width APDUs. If the `sign()` command does not return a fixed-width APDU, then indeed, this should be re-thought.

TL;DR: you may be right. This is news to me, but maybe we should be parsing

Reply:
Ok, I have removed the padding, and the ECDSA sig is left variable lenght... now to get the K81 driver to parse field types and lengths